File: CCC/2021/S3.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class S3 {
    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        int[] p = new int[n];
        int[] w = new int[n];
        int[] d = new int[n];

        int start = 1000000000;
        int end = 0;
        for (int i = 0; i < n; i++) {
            String[] s = reader.readLine().trim().split(" ");
            p[i] = Integer.parseInt(s[0]);
            start = Math.min(p[i], start);
            end = Math.max(p[i], end);
            w[i] = Integer.parseInt(s[1]);
            d[i] = Integer.parseInt(s[2]);
        }

        long result = 100000000000000000L;
        while (start <= end) {
            int middle = (start + end) / 2;
            long midr = walkTime(middle + 1, p, w, d);
            long midl = walkTime(middle - 1, p, w, d);
            long mid = walkTime(middle, p, w, d);
            result = Math.min(result, midr);
            result = Math.min(result, midl);
            result = Math.min(result, mid);
            if (midr < midl) {
                start = middle + 1;
            } else {
                end = middle - 1;
            }
        }

        System.out.println(result);
    }

    static long walkTime(int x, int[]p, int[] w, int[] d) {
        long time = 0;
        for (int i = 0; i < p.length; i++) {
            int rmax = p[i] + d[i];
            int lmax = p[i] - d[i];
            if (lmax <= x && x <= rmax) {
                continue;
            }
            int closest = lmax;
            if (x >= rmax) {
                closest = rmax;
            }

            time += (long) w[i] * Math.abs(x - closest);
        }

        return time;
    }
}

File: CCC/2021/S3.cpp
#include <bits/stdc++.h>

using namespace std;

long long walkTime(int x, vector<int> p, vector<int> w, vector<int> d);

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout << fixed;
    int n;
    cin >> n;
    vector<int> p(n);
    vector<int> w(n);
    vector<int> d(n);

    int start = 1000000000;
    int end = 0;

    for (int i = 0; i < n; i++)
    {
        cin >> p[i] >> w[i] >> d[i];
        start = min(p[i], start);
        end = max(p[i], end);
    }

    long long result = 100000000000000000;
    while (start <= end) {
        int middle = (start + end) / 2;
        long long midr = walkTime(middle + 1, p, w, d);
        long long midl = walkTime(middle - 1, p, w, d);
        long long mid = walkTime(middle, p, w, d);
        result = min(result, midr);
        result = min(result, midl);
        result = min(result, mid);
        if (midr < midl) {
            start = middle + 1;
        } else {
            end = middle - 1;
        }
    }

    cout << result;
}

long long walkTime(int x, vector<int> p, vector<int> w, vector<int> d) {
    long long time = 0;
    for (int i = 0; (size_t) i < p.size(); i++) {
        int rmax = p[i] + d[i];
        int lmax = p[i] - d[i];
        if (lmax <= x && x <= rmax) {
            continue;
        }
        int closest = lmax;
        if (x >= rmax) {
            closest = rmax;
        }

        time += (long long) w[i] * abs(x - closest);
    }

    return time;
}